core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala

```diff
@@ -0,0 +1,388 @@
+    }
+
+    // Unstash all messages stashed while in intermediate state
+    onTransition {
+        case _ -> Started => unstashAll()
+        case _ -> Ready   => unstashAll()
+        case _ -> Paused  => unstashAll()
+    }
+
+    initialize()
+
+    /** Delays all incoming messages until unstashAll() is called */
+    def delay = {
+        stash()
+        stay
+    }
+
+    /**
+     * Destroys the container after unpausing it if needed. Can be used
+     * as a state progression as it goes to Removing.
+     *
+     * @param container the container to destroy
+     */
+    def destroyContainer(container: Container) = {
+        context.parent ! ContainerRemoved
+
+        val unpause = stateName match {
+            case Paused => container.resume()(TransactionId.invokerNanny)
+            case _      => Future.successful(())
+        }
+
+        unpause
+            .flatMap(_ => container.destroy()(TransactionId.invokerNanny))
+            .map(_ => ContainerRemoved).pipeTo(self)
+
+        goto(Removing)
+    }
+
+    /**
+     * Runs the job, initialize first if necessary.
+     *
+     * @param container the container to run the job on
+     * @param job the job to run
+     * @return a future completing after logs have been collected and
+     *         added to the WhiskActivation
+     */
+    def run(container: Container, job: Run)(implicit tid: TransactionId): Future[WhiskActivation] = {
+        val actionTimeout = job.action.limits.timeout.duration
+
+        // Only initialize iff we haven't yet warmed the container
+        val initialize = stateData match {
+            case data: WarmedData => Future.successful(Interval.zero)
+            case _                => container.initialize(job.action.containerInitializer, actionTimeout)
+        }
+
+        val activation: Future[WhiskActivation] = initialize.flatMap { initInterval =>
+            val passedParameters = job.msg.content getOrElse JsObject()
+            val boundParameters = job.action.parameters.toJsObject
+            val parameters = JsObject(boundParameters.fields ++ passedParameters.fields)
+
+            val environment = JsObject(
+                "api_key" -> job.msg.user.authkey.compact.toJson,
+                "namespace" -> job.msg.user.namespace.toJson,
+                "action_name" -> job.msg.action.qualifiedNameWithLeadingSlash.toJson,
+                "activation_id" -> job.msg.activationId.toString.toJson,
+                // compute deadline on invoker side avoids discrepancies inside container
+                // but potentially under-estimates actual deadline
+                "deadline" -> (Instant.now.toEpochMilli + actionTimeout.toMillis).toString.toJson)
+
+            container.run(parameters, environment, actionTimeout)(job.msg.transid).map {
+                case (runInterval, response) =>
+                    val initRunInterval = Interval(runInterval.start.minusMillis(initInterval.duration.toMillis), runInterval.end)
```

Comment:
questioning again why `Interval.zero` actually need to know what instant of time it is "now". since this is only used to subtract from the run start time the duration of the initialization, having a duration of 0 regardless of now should be sufficient.

perhaps it's unavoidable with the current types but I don't see how the start time for a zero interval will matter (it will not reach the recover below, for example).

there is the case above where a "cold start" initialization failure also uses `zero` - but perhaps there is a way to use that interval constructor differently then in this case... and it could be that i'm just paying too much attention to a very small detail.